cpp/common/src/codingstandards/cpp/EncapsulatingFunctions.qll

```diff
@@ -18,6 +18,39 @@ class MainFunction extends MainLikeFunction {
   }
 }
 
+/**
+ * A test function from the GoogleTest infrastructure.
+ *
+ * Such functions can be treated as valid EntryPoint functions during analysis
+ * of "called" or "unused" functions. It is not straightforward to identify
+ * such functions, however, they have certain features that can be used for
+ * identification. This can be refined based on experiments/real-world use.
+ */
+class GTestFunction extends MainLikeFunction {
+  GTestFunction() {
+    // A GoogleTest function is named "TestBody" and
+    this.hasName("TestBody") and
+    // is enclosed by a class that inherits from a base class
+    this.getEnclosingAccessHolder() instanceof Class and
+    exists(Class base |
+      base = this.getEnclosingAccessHolder().(Class).getABaseClass() and
+      (
+        // called "Test" or
+        exists(Class c | base.getABaseClass() = c and c.hasName("Test"))
+        or
+        // defined under a namespace called "testing" or
+        exists(Namespace n | n = base.getNamespace() | n.hasName("testing"))
+        or
+        // is templatized by a parameter called "gtest_TypeParam_"
+        exists(TemplateParameter tp |
+          tp = base.getATemplateArgument() and
+          tp.hasName("gtest_TypeParam_")
+        )
```

Comment:
Can you provide an example of what this is capturing?

Reply:
I want to check if a function called `TestBody` is in a class that inherits a base class templatized by `gtest_TypeParam_`. If we look into the definitions of the [TYPED_TEST](https://github.com/google/googletest/blob/main/googletest/include/gtest/gtest-typed-test.h#L202) macro of GoogleTest, they use this specific name for the template parameter. We can also check it for the class that the `TestBody` member function belongs to, but since we already have a handle to the Base class, I used that. Let me know in case it can be improved.

Reply:
I am thinking of making this simpler. Kindly wait.

Reply:
Requesting to review commit: [a48c7cc](https://github.com/github/codeql-coding-standards/pull/725/commits/a48c7cc6af11ed445cee5a43d1e823db6b2c3833)